# EHI analysis
This repository contains the R code for analysing metagenomic data generated through the EHI pipeline.

The scripts are provided as RMarkdown documents. The tutorial-like information with test data is shown in the **html** documents rendered with knitr, while the original scripts are contained in the **Rmd** documents

## Analysis steps
1. **Environmenta preparation and data summary**: How to prepare the R environment for running the workflow, following by scripts to generate and visualise general statistics. html | [Rmd](EHI_analysis_1_data_summary.Rmd)
2. **MAG catalogue characterisation**: scripts to characterise the MAG catalogue
3. **Sequencing depth assessment**:
4. **Microbiome count data analysis**:
5. **Microbiome taxonomic composition analysis**:
6. **Microbiome diversity  analysis**:
